### PR TITLE
Split platform channel code into a separate package, flutter_channels

### PR DESCRIPTION
**Pull request by harryterkelsen**
## Description

This refactors the platform channel code into a separate package, package:flutter_channels. This is so plugins which are written in Dart (ie Web, Fuchsia plugins) can use the same Codec and Channel code as the framework.

## Related Issues

- https://github.com/flutter/flutter/issues/29629

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @dnfield has previously discussed moving all of foundation to a separate package.

cc @amirh for fuchsia plugins

File: .github/no-response.yml [hunk 1]
```diff
@@ -12,5 +12,5 @@ closeComment: >-
   resolve this issue. We are therefore reluctantly going to close this
   bug for now. Please don't hesitate to comment on the bug if you have
   any more information for us; we will reopen it right away!
-  
+
```
Comment (dnfield) [inline .github/no-response.yml]: I suspect GitHub needs this - and that you may not have meant to include this change.

Comment (hterkelsen) [reply .github/no-response.yml]: Removed this change. I originally added this because I was failing a test that was checking for trailing whitespace

**Comment by Hixie** [comment]
Is the plan to upload this package to pub?
- If so, then it shouldn't be in this repo.
- If not, then I don't understand the purpose of this change.

File: packages/flutter/test/services/platform_channel_test.dart [hunk 2]
```diff
@@ -231,10 +52,12 @@ void main() {
...
         'ch',
         (ByteData message) async {
-          final Map<dynamic, dynamic> methodCall = jsonMessage.decodeMessage(message);
+          final Map<dynamic, dynamic> methodCall =
+              jsonMessage.decodeMessage(message);
           if (methodCall['method'] == 'listen') {
             final String argument = methodCall['args'];
-            emitEvent(jsonMethod.encodeErrorEnvelope(code: '404', message: 'Not Found.', details: argument));
+            emitEvent(jsonMethod.encodeErrorEnvelope(
+                code: '404', message: 'Not Found.', details: argument));
```
Comment (Hixie) [inline packages/flutter/test/services/platform_channel_test.dart]: please don't make changes like this, this wrapping style violates our style guide

File: packages/flutter_channels/lib/src/binary_messenger.dart [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (Hixie) [inline packages/flutter_channels/lib/src/binary_messenger.dart]: this copyright date is invalid, this code is much older than that.

File: packages/flutter_channels/lib/src/binary_messenger.dart [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  /// Set a callback for receiving messages from the platform plugins on the
+  /// given channel, without decoding them.
+  ///
+  /// The given callback will replace the currently registered callback for that
+  /// channel, if any. To remove the handler, pass null as the [handler]
+  /// argument.
+  ///
+  /// The handler's return value, if non-null, is sent as a response, unencoded.
+  void setMessageHandler(
+      String channel, Future<ByteData> handler(ByteData message));
```
Comment (Hixie) [inline packages/flutter_channels/lib/src/binary_messenger.dart]: please use flutter style here and elsewhere

**Comment by hterkelsen** [comment]
I'm abandoning this PR in favor of https://github.com/flutter/flutter/pull/30406

**Pull request closed without merging**